### ISSUE TITLE
Collaboration

### ISSUE THREAD

**Issue by tpluscode** [vid/shacl-poc#1]
Hi. I found your repo via your usage `@tpluscode/rdf-ns-builders`

It so happens that I have a near fully-featured SHACL form builder: https://forms.hypermedia.app

Would you be interested in joining efforts?

**Comment by vid** [comment]
Yes! I stumbled across shaperone recently and meant to follow up with regard to support for form validation and logical constraints. I'd be more than happy to collaborate if it meets goals. Maybe a chat? Let me know what works. Thanks.

**Comment by tpluscode** [comment]
Totally. We can join a call some time and see how much we align. I'm sure we can work something out :). Hit me on awesome@hypermedia.app

**Comment by vid** [comment]
@tpluscode , thanks for the discussion and follow up. 

This shacl-poc repository is working on a subset of SHACL capabilities for limited use cases. For fully realizing the capabilities of SHACL we would look to transition to a project like shaperone. However, since the POC is very limited and specific, I am not certain we could use shaperone today.

I can't really evaluate all the nuance of SHACL and your implementation, so can't evaluate where our use case may not be supported. I understand there may be some ambiguity using logical constraints, as discussed in https://github.com/hypermedia-app/shaperone/issues/117, however for now, only using xone, it _seems_  straightforward to allow the user to choose between the two and only use that in the data graph/validation.

Regarding displaying validation results inline, https://github.com/hypermedia-app/shaperone/issues/118, since we are using your library we may encounter this problem (the thought, without digging into the implementation, was blank nodes identifiers might provide a solution?).

Resources are really limited to develop the poc (and there are a few things to do outside the form). With a clear example addressing both requirements the poc can switch to your library, however otherwise the focus has to be conformant SHACL and a built for purpose library that can be switched out in the future.

We talked about some code to collaborate on this, maybe the shaperone playground can be used to try these features?

I hope this makes sense!

**Comment by tpluscode** [comment]
Certainly, the playground can be used to experiment within the limits of the implemented features. Additionally, any PR deploys a playground with its changes to try out. You can [use it already](https://logical-constraints--shaperone.netlify.app/playground/?resource=%7B%0A++%22%40context%22%3A+%7B%0A++++%22rdf%22%3A+%22http%3A%2F%2Fwww.w3.org%2F1999%2F02%2F22-rdf-syntax-ns%23%22%2C%0A++++%22rdfs%22%3A+%22http%3A%2F%2Fwww.w3.org%2F2000%2F01%2Frdf-schema%23%22%2C%0A++++%22xsd%22%3A+%22http%3A%2F%2Fwww.w3.org%2F2001%2FXMLSchema%23%22%2C%0A++++%22schema%22%3A+%22http%3A%2F%2Fschema.org%2F%22%0A++%7D%2C%0A++%22%40id%22%3A+%22http%3A%2F%2Fexample.com%2FJohn_Doe%22%2C%0A++%22%40type%22%3A+%22schema%3APerson%22%2C%0A++%22schema%3AfamilyName%22%3A+%22Doe%22%2C%0A++%22schema%3AfirstName%22%3A+%22John%22%0A%7D&selectedResource=http%3A%2F%2Fexample.com%2FJohn_Doe&shapes=%40prefix+sh%3A+%3Chttp%3A%2F%2Fwww.w3.org%2Fns%2Fshacl%23%3E+.%0A%40prefix+schema%3A+%3Chttp%3A%2F%2Fschema.org%2F%3E+.%0A%40prefix+rdfs%3A+%3Chttp%3A%2F%2Fwww.w3.org%2F2000%2F01%2Frdf-schema%23%3E+.%0A%40prefix+ex%3A+%3Chttp%3A%2F%2Fexample.com%2F%3E+.%0A%0Aex%3APersonShape%0A++a+sh%3AShape+%3B%0A++sh%3AtargetClass+schema%3APerson+%3B%0A++rdfs%3Alabel+%22Person%22+%3B%0A++sh%3Aproperty+ex%3ALastNameProperty+%3B%0A++sh%3Aor+%28+ex%3AWithFirstName+ex%3AWithGivenName+%29%0A.%0A%0Aex%3ALastNameProperty+%0A++a+sh%3APropertyShape+%3B%0A++sh%3Apath+schema%3AfamilyName+%3B%0A++sh%3Aorder+20+%3B%0A++sh%3AmaxCount+1+%3B%0A.%0A%0Aex%3AWithFirstName%0A++sh%3Aproperty+%5B%0A++++sh%3Apath+schema%3AfirstName+%3B++++%0A++++sh%3Aorder+10+%3B%0A++++sh%3AmaxCount+1+%3B%0A++%5D+%3B%0A.%0A%0Aex%3AWithGivenName%0A++sh%3Aproperty+%5B%0A++++sh%3Apath+schema%3AgivenName+%3B%0A++++sh%3Aorder+10+%3B%0A++++sh%3AmaxCount+1+%3B%0A++%5D+%3B%0A.&resourcePrefixes=schema&disableEditorChoice=true) to see the current extent of hypermedia-app/shaperone#117. As discussed in the related issue and PR, this will simply combine all properties to at least have a complete UI representing all the fields.

> [...] only using xone, it _seems_ straightforward to allow the user to choose between the two [...]

This may seem easy when there are only two fields, and if you start from an empty form. But when you add more properties to the alternative shapes or the data graph is nonconformant to begin with, you would easily end up with a rendering where some hidden properties generate validation errors, which would be confusing to the users.

Shaperone aims to be a generic tool, an I'm not yet confident in how these features would be implemented in a way to satisfy different use cases.

That said, the rendering can be customised to accommodate different choices in the actual rendered output. (I now realise that page is missing from the documentation, should work on that soon too)

> [...] and only use that in the data graph/validation.

This is not how validation would work. You would not split the shape but rather pass it to a validation library along the data graph to get the report.

> Resources are really limited to develop the poc [...]

It is up to you. I can offer my help in adding features to Shaperone and/or helping you build your custom extensions where necessary and I won't feel ready to introduce them to the core. Between the two of us, this might be a more efficient way to achieve your goal. And soon you would be capable of contributing yourself.

> otherwise the focus has to be conformant SHACL 

I'm a little lost here. "Conformant SHACL" as opposed to what? 

> Regarding displaying validation results inline

Regarding validation, the hard part is not really presenting errors but actually producing a useful report. The [rdf-validate-shacl](https://npm.im/rdf-validate-shacl) library we're currently using has limitations in the level of detail and effort would have to shift there.

In Shaperone, the implementation would be agnostic of how you perform validation and only follow the validation report spec to connect messages to fields and to the rendering.

**Comment by vid** [comment]
By "Conformant SHACL," I just meant to emphasize that we want it to be forward-compatible with other libraries.

I would like to use Shaperone (with Hydra) but given we have an April timeline, it's a limited project, and its requirements aren't currently a priority for Shaperone, will have to continue with our limited approach. I'm switching my focus to this dev later this week, will look for opportunities to contribute to Shaperone. Thank you very much for your patient explanations.

**Comment by tpluscode** [comment]
Well that's a shame, but hopefully you will reconsider at some point.

At any rate, all the features you need are definitely within scope for a future 1.0 of shaperone, even if we did not need them immediately in our project.

I have also started working on validation and [here's a example of its current state](https://validation-report--shaperone.netlify.app/playground/?resource=%7B%0A++%22%40context%22%3A+%7B%0A++++%22rdf%22%3A+%22http%3A%2F%2Fwww.w3.org%2F1999%2F02%2F22-rdf-syntax-ns%23%22%2C%0A++++%22rdfs%22%3A+%22http%3A%2F%2Fwww.w3.org%2F2000%2F01%2Frdf-schema%23%22%2C%0A++++%22xsd%22%3A+%22http%3A%2F%2Fwww.w3.org%2F2001%2FXMLSchema%23%22%2C%0A++++%22schema%22%3A+%22http%3A%2F%2Fschema.org%2F%22%2C%0A++++%22foaf%22%3A+%22http%3A%2F%2Fxmlns.com%2Ffoaf%2F0.1%2F%22%2C%0A++++%22vcard%22%3A+%22http%3A%2F%2Fwww.w3.org%2F2006%2Fvcard%2Fns%23%22%0A++%7D%2C%0A++%22%40graph%22%3A+%5B%0A++++%7B%0A++++++%22%40id%22%3A+%22http%3A%2F%2Fexample.com%2FJane_Doe%22%2C%0A++++++%22%40type%22%3A+%22schema%3APerson%22%2C%0A++++++%22schema%3Aname%22%3A+%22Janet%22%0A++++%7D%2C%0A++++%7B%0A++++++%22%40id%22%3A+%22http%3A%2F%2Fexample.com%2FJohn_Doe%22%2C%0A++++++%22%40type%22%3A+%22schema%3APerson%22%2C%0A++++++%22schema%3Aage%22%3A+%7B%0A++++++++%22%40type%22%3A+%22xsd%3Ainteger%22%2C%0A++++++++%22%40value%22%3A+%2217%22%0A++++++%7D%2C%0A++++++%22schema%3Aknows%22%3A+%7B%0A++++++++%22%40id%22%3A+%22http%3A%2F%2Fexample.com%2FJane_Doe%22%0A++++++%7D%2C%0A++++++%22schema%3Aname%22%3A+%22John+Doe%22%2C%0A++++++%22foaf%3Agender%22%3A+%22Malee%22%0A++++%7D%0A++%5D%0A%7D&selectedResource=http%3A%2F%2Fexample.com%2FJohn_Doe&shapes=%40prefix+sh%3A+%3Chttp%3A%2F%2Fwww.w3.org%2Fns%2Fshacl%23%3E+.%0A%40prefix+schema%3A+%3Chttp%3A%2F%2Fschema.org%2F%3E+.%0A%40prefix+rdfs%3A+%3Chttp%3A%2F%2Fwww.w3.org%2F2000%2F01%2Frdf-schema%23%3E+.%0A%40prefix+xsd%3A+%3Chttp%3A%2F%2Fwww.w3.org%2F2001%2FXMLSchema%23%3E+.%0A%40prefix+dash%3A+%3Chttp%3A%2F%2Fdatashapes.org%2Fdash%23%3E+.%0A%40prefix+foaf%3A+%3Chttp%3A%2F%2Fxmlns.com%2Ffoaf%2F0.1%2F%3E+.%0A%40prefix+vcard%3A+%3Chttp%3A%2F%2Fwww.w3.org%2F2006%2Fvcard%2Fns%23%3E+.%0A%0A%40prefix+ex%3A+%3Chttp%3A%2F%2Fexample.com%2F%3E+.%0A%40prefix+lexvo%3A+%3Chttp%3A%2F%2Flexvo.org%2Fid%2Fiso639-1%2F%3E+.%0A%0Aex%3APersonShape%0A++a+sh%3AShape+%3B%0A++sh%3AtargetClass+schema%3APerson+%3B%0A++rdfs%3Alabel+%22Person%22+%3B%0A++sh%3Aproperty+ex%3ANameProperty+%2C%0A+++++++++++++++++ex%3AKnowsProperty+%2C%0A+++++++++++++++++ex%3AAgeProperty+%2C%0A+++++++++++++++++ex%3AGenderProperty+%2C%0A+++++++++++++++++ex%3ASpokenLanguagesProperty++%2C%0A+++++++++++++++++ex%3ADateOfBirthProperty+%2C%0A+++++++++++++++++ex%3AHomePageProperty+%3B%0A.%0A%0Aex%3ASimplifiedPersonShape%0A++a+sh%3AShape+%3B%0A++sh%3AtargetNode+ex%3AJane_Doe+%3B%0A++rdfs%3Alabel+%22Person+%28name-only%29%22+%3B%0A++sh%3Aproperty+ex%3ANameProperty+%3B%0A.%0A%0Aex%3ANameProperty%0A++sh%3Apath+schema%3Aname+%3B%0A++sh%3Aname+%22Name%22+%3B%0A++sh%3Adatatype+xsd%3Astring+%3B%0A++dash%3AsingleLine+true+%3B%0A++sh%3AmaxCount+1+%3B%0A++sh%3AminCount+1+%3B%0A++sh%3Aorder+1+%3B%0A.%0A%0Aex%3AKnowsProperty%0A++sh%3Apath+schema%3Aknows+%3B%0A++sh%3Aclass+schema%3APerson+%3B%0A++sh%3Agroup+ex%3AFriendGroup+%3B%0A.%0A%0Aex%3AAgeProperty%0A++sh%3Apath+schema%3Aage+%3B%0A++sh%3Aname+%22Age%22+%3B%0A++sh%3Adatatype+xsd%3Ainteger+%3B%0A++sh%3AmaxCount+1+%3B%0A++sh%3AdefaultValue+21+%3B%0A++sh%3Aorder+2+%3B%0A++sh%3AminInclusive+18+%3B%0A.%0A%0Aex%3AGenderProperty%0A++sh%3Apath+foaf%3Agender+%3B%0A++sh%3Aname+%22Gender%22+%3B%0A++sh%3Ain+%28%0A++++%22Male%22+%22Female%22+%22Other%22+%22Prefer+not+to+tell%22%0A++%29+%3B%0A++sh%3AmaxCount+1+%3B%0A++sh%3Aorder+3+%3B%0A++sh%3Amessage+%22Please+select+a+valid+gender%22+%3B%0A.%0A%0Aex%3ADateOfBirthProperty%0A++sh%3Apath+schema%3AbirthDate+%3B%0A++sh%3Aname+%22Date+of+birth%22+%3B%0A++sh%3AmaxCount+1+%3B%0A++sh%3Aorder+4+%3B%0A++sh%3Adatatype+xsd%3Adate+%3B%0A.%0A%0Aex%3ASpokenLanguagesProperty%0A++sh%3Apath+vcard%3Alanguage+%3B%0A++sh%3Aname+%22Spoken+languages%22+%3B%0A++sh%3AnodeKind+sh%3AIRI+%3B%0A++sh%3Ain+%28%0A++++lexvo%3Aen+lexvo%3Ade+lexvo%3Afr+lexvo%3Apl+lexvo%3Aes%0A++%29+%3B%0A++sh%3Aorder+5+%3B%0A++sh%3AminCount+1+%3B%0A++sh%3AmaxCount+2+%3B%0A.%0A%0Aex%3AHomePageProperty%0A++sh%3Apath+foaf%3Ahomepage+%3B%0A++sh%3Aname+%22Homepage+URL%22+%3B%0A++sh%3AnodeKind+sh%3AIRI+%3B%0A++sh%3Aorder+6+%3B%0A.%0A%0Aex%3AFriendGroup%0A++a+sh%3APropertyGroup+%3B%0A++rdfs%3Alabel+%22Acquaintances%22%0A.%0A%0Alexvo%3Aen+rdfs%3Alabel+%22English%22+.%0Alexvo%3Ade+rdfs%3Alabel+%22German%22+.%0Alexvo%3Afr+rdfs%3Alabel+%22French%22+.%0Alexvo%3Apl+rdfs%3Alabel+%22Polish%22+.%0Alexvo%3Aes+rdfs%3Alabel+%22Spanish%22+.&components=material), built from https://github.com/hypermedia-app/shaperone/pull/122

**Comment by vid** [comment]
The validation example looks great. In terms of UX, the main point that may be confusing is seeing both the count and type errors, the later of which is not very user friendly.

![image](https://user-images.githubusercontent.com/68174/113150029-a4a79200-9201-11eb-9ae9-be56de4375aa.png)

As for use of Shaperone, believe me, it's not a matter of considering, since it's clearly the better path, it's just a matter of current fit and the scope of the poc.

Taking it one step at a time, if Shaperone could present a simplified view for `xone`, we could think about working with it for the current development. I think for many situations `xone` will be the right choice, it certainly seems like the easiest to work with. 

For example, if there is only one type of logical constraint, and it's `xone`, provide a tabbed interface for input and only include that value node in the graph. However, maybe a special case for "only uses xone" doesn't suit Shaperone as a general purpose library.

**Comment by tpluscode** [comment]
Surely, this is but a first attempt at displaying the errors. I will take any suggestions for improvement :)

The errors come from an external library. Maybe they could be customised or modified before displaying, at least by shortening the URL so that it at least becomes `Value does not have datatype xsd:integer`

> a simplified view for `xone`

The core library should provide some more metadata about properties, including the logical constraint, if any.

That way a customisation would be easily made to the rendering, such as hiding or disabling the excluded alternatives. I'll be happy to provide an example

**Comment by vid** [comment]
> Surely, this is but a first attempt at displaying the errors. I will take any suggestions for improvement :)

I realize that, in fact I saw it change between reloads. I think the approach of the error message inline rather than a disappearing tooltip is better. 

> I'll be happy to provide an example

That would be really helpful. In case it's not clear, everyone would much prefer to use Shaperone and help develop it (even if just on the edges).

**Comment by tpluscode** [comment]
In fact, the disappearing tooltip is html behaviour of native inputs. Proper design system would look better, I'm sure

**Comment by saumier** [comment]
Hi. This is a really interesting discussion :-)  Looks like Shaperone is very close to being a perfect fit for our projects at Culture Creates.
